plugin/src/main/java/org/opensearch/ml/processor/MLInferenceSearchResponseProcessor.java

```diff
@@ -144,20 +153,107 @@ public void processResponseAsync(
+                    response,
+                    responseListener,
+                    hits
+                );
+
+                for (SearchHit hit : hits) {
+                    SearchHit[] newHits = new SearchHit[1];
+                    newHits[0] = hit;
+                    SearchResponse oneHitResponse = SearchResponseUtil.replaceHits(newHits, response);
+                    ActionListener<SearchResponse> oneHitListener = getOneHitListener(combineResponseListener);
+                    rewriteResponseDocuments(oneHitResponse, oneHitListener);
+                }
+            }
+
         } catch (Exception e) {
             if (ignoreFailure) {
                 responseListener.onResponse(response);
             } else {
-                responseListener.onFailure(e);
+                responseListener
+                    .onFailure(new OpenSearchStatusException("Failed to process response: " + e.getMessage(), RestStatus.BAD_REQUEST));
             }
         }
     }
 
+    /**
+     * Creates an ActionListener for a single SearchResponse that delegates its
+     * onResponse and onFailure callbacks to a GroupedActionListener.
+     *
+     * @param combineResponseListener The GroupedActionListener to which the
+     *                                 onResponse and onFailure callbacks will be
+     *                                 delegated.
+     * @return An ActionListener that delegates its callbacks to the provided
+     *         GroupedActionListener.
+     */
+    private static ActionListener<SearchResponse> getOneHitListener(GroupedActionListener<SearchResponse> combineResponseListener) {
+        ActionListener<SearchResponse> oneHitListener = new ActionListener<>() {
+            @Override
+            public void onResponse(SearchResponse response) {
+                combineResponseListener.onResponse(response);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                combineResponseListener.onFailure(e);
+            }
+        };
+        return oneHitListener;
+    }
+
+    /**
+     * Creates a GroupedActionListener that combines the SearchResponses from individual hits
+     * and constructs a new SearchResponse with the combined hits.
+     *
+     * @param response         The original SearchResponse containing the hits to be processed.
+     * @param responseListener The ActionListener to be notified with the combined SearchResponse.
+     * @param hits             The array of SearchHits to be processed.
+     * @return A GroupedActionListener that combines the SearchResponses and constructs a new SearchResponse.
+     */
+    private GroupedActionListener<SearchResponse> getCombineResponseGroupedActionListener(
+        SearchResponse response,
+        ActionListener<SearchResponse> responseListener,
+        SearchHit[] hits
+    ) {
+        GroupedActionListener<SearchResponse> combineResponseListener = new GroupedActionListener<>(new ActionListener<>() {
```

Comment:
Ensure that `GroupedActionListener` is thread-safe, as it appears to be aggregating responses from potentially asynchronous operations. If multiple threads call onResponse or onFailure concurrently, this could lead to race conditions.

Reply:
right. this groupedActionListener collects response from multiple oneHitListener but it would onResponse after all  multiple oneHitListener returns.